src/ejabberd_c2s.erl

```diff
@@ -737,41 +737,33 @@ do_open_session(Acc, JID, StateData) ->
             end
     end.
 
-do_open_session_common(Acc, JID, #state{user = U, resource = R} = NewStateData0) ->
+do_open_session_common(Acc, JID, #state{user = U, server = S, resource = R} = NewStateData0) ->
     change_shaper(NewStateData0, JID),
-    Acc1 = ejabberd_hooks:run_fold(roster_get_subscription_lists,
-                                  NewStateData0#state.server,
-                                  Acc,
-                                  [U, NewStateData0#state.server]),
+    Acc1 = ejabberd_hooks:run_fold(roster_get_subscription_lists, S, Acc, [U, S]),
     {Fs, Ts, Pending} = mongoose_acc:get(roster, subscription_lists, {[], [], []}, Acc1),
     LJID = jid:to_lower(jid:to_bare(JID)),
     Fs1 = [LJID | Fs],
     Ts1 = [LJID | Ts],
-    PrivList = ejabberd_hooks:run_fold(privacy_get_user_list,
-                                       NewStateData0#state.server,
-                                       #userlist{},
-                                       [U, NewStateData0#state.server]),
+    PrivList = ejabberd_hooks:run_fold(privacy_get_user_list, S, #userlist{}, [U, S]),
     SID = {p1_time_compat:timestamp(), self()},
     Conn = get_conn_type(NewStateData0),
     Info = [{ip, NewStateData0#state.ip}, {conn, Conn},
             {auth_module, NewStateData0#state.auth_module}],
-    ReplacedPids = ejabberd_sm:open_session(SID, U, NewStateData0#state.server, R, Info),
+    ReplacedPids = ejabberd_sm:open_session(SID, U, S, R, Info),
 
-    MonitorRefs = ordsets:from_list([{monitor(process, PID), PID} || PID <- ReplacedPids]),
-    lists:foreach(
-        fun({MonitorRef, PID}) ->
-            receive
```

Comment:
This selective receive makes `do_open_session_common` a transaction. It writes to the `session` table and waits for all previous resources to terminate, before it starts processing any other messages from the process's inbox. This is not accidental behaviour, it's intentional.

Are we certain that we can allow processing other messages (if any) between handling `DOWN` signals? In other words, is it safe to move the receive outside this functions and rely on `handle_info`?

Reply:
In a version from current master there is a window between opening a new session and waiting for old one(s) to close anyway. If we make the synchronous wait too long - it will impact UX. Too short results in false warnings about unknown `DOWN` messages. I can't think of any message that could be harmful here. This `receive` is a poor man's protection anyway, because it doesn't e.g. force kill old session after timeout.
New C2S will reject any replies meant only for old session thanks to `check_incoming_accum_for_conflicts`.

Reply:
Ok, sounds good!